src/gpodder/libgpod_ctypes.py

```diff
@@ -0,0 +1,428 @@
+#
+# -*- coding: utf-8 -*-
+#
+# gPodder - A media aggregator and podcast client
+# Copyright (c) 2005-2022 The gPodder Team
+#
+# gPodder is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 3 of the License, or
+# (at your option) any later version.
+#
+# gPodder is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+
+#
+# libgpod_ctypes: Minimalistic ctypes-based bindings for libgpod
+# (Just enough coverage to get podcast syncing working again...)
+# Thomas Perl <m@thp.io>, May 2022
+#
+
+
+import ctypes
+import logging
+import os
+import struct
+
+logger = logging.getLogger(__name__)
+
+
+# libgpod, for iTunesDB access
+libgpod = ctypes.CDLL('libgpod.so.4')
+
+# glib, for g_strdup() and g_free()
+libglib = ctypes.CDLL('libglib-2.0.so.0')
+
+
+# glib/gtypes.h: typedef gint   gboolean;
+gboolean = ctypes.c_int
+
+# glib/gstrfuncs.h: gchar *g_strdup(const gchar *str);
+libglib.g_strdup.argtypes = (ctypes.c_char_p,)
+# Note: This MUST be c_void_p, so that the glib-allocated buffer will
+# be preserved when assigning to track member variables. The reason
+# for this is that Python ctypes tries to be helpful and converts a
+# c_char_p restype to a Python bytes object, which will be different
+# from the memory returned by g_strdup(). For track properties, the
+# values will be free'd indirectly by itdb_free() later.
+libglib.g_strdup.restype = ctypes.c_void_p
+
+# glib/gmem.h: void g_free(gpointer mem);
+libglib.g_free.argtypes = (ctypes.c_void_p,)
+libglib.g_free.restype = None
+
+# XXX: Can we always assume time_t is 64-bit these days?
```

Comment:
You can assume my 32bit system is 64bit, but it would be incorrect.

`platform.architecture()[0]` returns '32bit' or '64bit' and could be used to change the size.

Reply:
Yes and no. See also: https://linux.slashdot.org/story/20/02/15/0247201/linux-is-ready-for-the-end-of-time

I haven't checked with different libc versions what the size of time_t is.

Related glibc documentation: https://www.gnu.org/software/libc/manual/html_node/64_002dbit-time-symbol-handling.html

Just using pointer size for this is probably wrong (but so might be assuming it's 64-bit always, hence the added comment...).

Reply:
Well, time_t **is** 32bit on my 32bit system. I have no plans to use libgpod (no ipod), so we could just let it crash or do bad things for anyone with a 32bit time_t. Maybe at least detect a 32bit platform and log a message that libgpod support might not work if time_t is 32bit.

Reply:
I'll change it to assume 32-bit time_t on 32-bit platforms, but while looking for a way to maybe "find" the time_t size somewhere in Python, I spotted this in the CPython source code:

```
        /* MS VS2005 changes time_t to a 64-bit type on all platforms */
#       if defined(_MSC_VER) && _MSC_VER >= 1400
#       define SIZEOF_TIME_T 8
#       else
#       define SIZEOF_TIME_T 4
#       endif
```

https://github.com/python/cpython/blob/d284e8b3e351c69406519d5afdcdbed7fbbe0190/PC/pyconfig.h#L301

Reply:
Looks good, but I still think the 32bit path should log a message to indicate that time_t might be 64bit and might causes issues. Like you mentioned below, there are 32bit platforms that use 64bit time_t, and I read that musl on linux uses it. Being able to ask users if the message exists would be helpful for any issues they open about it.

```python
    logger.info('libgpod may cause issues if time_t is 64-bit on your 32-bit system.')
```